contracts/vaults/VaultPauser/RibbonVaultPauser.sol

```diff
@@ -0,0 +1,253 @@
+
+        PauseReceipt[] storage pauseReceipts =
+            pausedPositions[currentVaultAddress][_account];
+
+        // user can pause multiple position hence it's necessary to make an array
+        // to store each paused position
+        pauseReceipts.push(
+            PauseReceipt({
+                round: uint16(currentVault.vaultState().round),
+                account: address(_account),
+                shares: uint104(_amount)
+            })
+        );
+
+        emit Pause(
+            _account,
+            currentVaultAddress,
+            _amount,
+            currentVault.vaultState().round
+        );
+
+        currentVault.initiateWithdraw(_amount);
+    }
+
+    /**
+     * @notice resume user's position into vault by making a deposit
+     * @param _vaultAddress vault's address
+     */
+    function resumePosition(address _vaultAddress) external override {
+        IRibbonThetaVault currentVault = IRibbonThetaVault(_vaultAddress);
+
+        address currentUser = address(msg.sender);
+
+        PauseReceipt[] memory pauseReceipts =
+            pausedPositions[_vaultAddress][currentUser];
+
+        // loop all receipts, convert into withdraw amount and sum it up
+        uint256 totalWithdrawAmount = 0;
+        uint256 receiptsLength = pauseReceipts.length;
+        for (uint16 i = 0; i < receiptsLength; i++) {
+            if (pauseReceipts[i].round < currentVault.vaultState().round) {
+                totalWithdrawAmount += ShareMath.sharesToAsset(
+                    pauseReceipts[i].shares,
+                    currentVault.roundPricePerShare(
+                        uint256(pauseReceipts[i].round)
+                    ),
+                    currentVault.vaultParams().decimals
+                );
+            }
+        }
+
+        // delete receipts once finish calculating total withdraw amount
+        delete pausedPositions[_vaultAddress][currentUser];
+
+        string memory currentSymbol = currentVault.symbol();
+
+        // stETH transfers suffer from an off-by-1 error
+        // since we received STETH , we shall deposit using STETH instead of ETH
+        if (
+            (keccak256(abi.encodePacked(currentSymbol)) ==
+                keccak256(abi.encodePacked("rSTETH-THETA")))
+        ) {
+            totalWithdrawAmount = totalWithdrawAmount.sub((3 * receiptsLength));
+
+            emit Resume(currentUser, _vaultAddress, totalWithdrawAmount.sub(1));
+            IERC20(STETH).approve(_vaultAddress, totalWithdrawAmount);
+            currentVault.depositYieldToken(totalWithdrawAmount, currentUser);
+            return;
+        }
+
+        emit Resume(currentUser, _vaultAddress, totalWithdrawAmount);
+        // if asset is ETH, we will convert it into WETH before depositing
+        if (currentVault.vaultParams().asset == WETH) {
+            IWETH(WETH).deposit{value: totalWithdrawAmount}();
+        }
+        IERC20(currentVault.vaultParams().asset).approve(
```

Comment:
Each call to `currentVault.vaultParams().asset` costs a lot of gas since they access storage variables. You will have to cache them into a variable

```solidity
address asset = currentVault.vaultParams().asset
```